src/main/java/com/netflix/playback/features/model/impl/DiagnosticServiceImpl.java

```diff
@@ -0,0 +1,87 @@
+package com.netflix.playback.features.model.impl;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import com.netflix.playback.features.model.DiagnosticService;
+import com.netflix.playback.features.model.Timer;
+/**
+ *  The service maintains the instance of timer and start event of it
+ *  The service stores the log events for the current time period.
+ *  When timer is stopped the callback is invoked and the service flushes the data to aggregator 
+ *  for various calculations.
+ */
+public class DiagnosticServiceImpl extends DiagnosticService {
+
+	//store the events count for the current time period
+	private final ConcurrentMap<Integer, AtomicInteger> currentPeriodCount = new ConcurrentHashMap<>();
+	private final PeriodAggregator aggregator = new PeriodAggregator();
+	private final Lock lock = new ReentrantLock();
+	 
+	public DiagnosticServiceImpl(Timer timer, int numCompletedPeriods) {
+		super(timer, numCompletedPeriods);
+	}
+
+	@Override
+	public int rate(int key) {
+		Optional<Integer> lastRateOptional = aggregator.getRateForPeriod(key, numCompletedPeriods);
+		Optional<Integer> lastMinusOneRateOptional = aggregator.getRateForPeriod(key, numCompletedPeriods -1);
+		if (lastRateOptional.isPresent() && lastMinusOneRateOptional.isPresent()) {
+			return (lastRateOptional.get() - lastMinusOneRateOptional.get());
+		} else if (lastRateOptional.isPresent()) {
+			return lastRateOptional.get();
+		} else {
+			return lastMinusOneRateOptional.get() * -1;
```

Comment:
so if lastRate is not present but lastMinusOne is present, i am calculating diff as ( 0 - lastMinusOneRateOptional.get()) which is lastMinusOneRateOptional.get() * -1; 
there should be an if check on line 37 too.
May be we can just return 0 if the lastRate is not present